MCO-52: add MCO API into openshift/api

jkyros opened the pull request:
This is an [existing API](https://github.com/openshift/machine-config-operator/tree/master/pkg/apis/machineconfiguration.openshift.io), so fortunately or unfortunately we're probably stuck with some of the...tech debt here :frowning_face: 

We've tried this before, and while there wasn't really anything we were "hard blocked" on, it never quite got finished. 

Iif you want to re-read previous attempts/reviews: 

- https://github.com/openshift/api/issues/216 (historical protobuf context, not an issue anymore because we use RawExtension for ignition now)
-  https://github.com/openshift/machine-config-operator/issues/588
- https://github.com/openshift/api/pull/301
-  https://github.com/openshift/api/pull/620
- https://github.com/openshift/api/pull/1040
-  https://github.com/openshift/api/pull/1060

This time is going to be the one though, I can feel it! :smile:  

After this is done, we'll need to do client-go and pull it back into the MCO: 
- https://github.com/openshift/client-go/pull/243
- https://github.com/openshift/machine-config-operator/pull/3747

What I did: 
- took the existing MCO API from https://github.com/openshift/machine-config-operator/tree/master/pkg/apis/machineconfiguration.openshift.io
- added protobuf tags 
- added an operator CRD (`machineconfigurations.operator.openshift.io`) for the MCO since it did not have one  (this should let us [unblock our move](https://github.com/openshift/machine-config-operator/pull/2882) away from our library-go fork, since we'll have a place to store our generations) 
- added compatibility level annotations
- added more detailed field descriptions in some cases

jkyros commented on machineconfiguration/v1/helpers.go:
```diff
@@ -0,0 +1,199 @@
+package v1
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NewMachineConfigPoolCondition creates a new MachineConfigPool condition.
+func NewMachineConfigPoolCondition(condType MachineConfigPoolConditionType, status corev1.ConditionStatus, reason, message string) *MachineConfigPoolCondition {
```
These helpers aren't used anywhere but in the MCO to my knowledge. I assume you'll want them excluded, but I figured there was no harm in asking? :smile:

jkyros commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,654 @@
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +openshift:compatibility-gen:internal
+type ControllerConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// TODO(jkyros): inconsistent historical generation resulted in the controllerconfig CRD being
+	// generated with all fields required, while everything else was generated with optional
+
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ControllerConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```
Hmm that failed integration test made me poke through all the required fields. 

The MCO hasn't been historically great at required/non-required fields since most of our CRD updates had been done manually.  It looks like ControllerConfig was originally (years ago) generated with "required by default" at the package level and has been updated that way, while everything else was generated with "optional by default" at the package level and maintained that way.  

This PR initially defaulted to "fields required by default" at the package level. Looking at the rest of the APIs in the repo, it looks like there's a mix?

I've adjusted this now so that:
-  `// +kubebuilder:validation:Optional` is set at the package level 
- the `ControllerConfig` fields that were previously required are set explicitly to required 
- the tests should be fixed and now supply the required fields (I will clean the data up before merge) 

Generally though, what are your "best practice" thoughts on how required fields should be handled: 
- Required by default, mark everything we don't want required as optional or 
- Optional by default, mark everything we don't want optional as required 
- Mark literally every field explicitly so it doesn't matter? :smile:

deads2k commented:
/hold

We definitely want this, but we have some other churn in progress that must complete first.  Then we take this to reduce future debt in 4.14.

jkyros commented:
Rebased, set our runlevel to 80 to match existing MCO, adjusted approval annotation to match this PR, pulled out the helpers :smile:

jkyros commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,654 @@
+	// +required
+	Spec ControllerConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// +optional
+	Status ControllerConfigStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
```
The MCO itself does not care about the protobufs or use them, so I'm cool with removing them.

Previous attempts at migration seemed to really care a lot about getting them in there but the reason was unclear. It could have been either cargo-cult or for something that has since gone away? 

I just want to make sure we don't have some important thing somewhere that's needs protobufs that I don't know about. 

Should I take them out?

JoelSpeed replied:
Protobufs have no effect for CRD types, so they should be removed, please make that its own commit

jkyros commented:
Rebased and trued up with recent MCO merges. I'll squash the extra commits out when we're cool with this.

jkyros commented on operator/v1/register.go:
```diff
@@ -54,6 +54,8 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&KubeSchedulerList{},
 		&KubeStorageVersionMigrator{},
 		&KubeStorageVersionMigratorList{},
+		&MachineConfiguration{},
+		&MachineConfigurationList{},
```
Just added MachineConfiguration to operator register.go so @cdoern  could get client-go to generate properly

JoelSpeed commented on machineconfiguration/v1/0000_80_containerruntimeconfig.crd.yaml:
```diff
@@ -113,10 +113,6 @@ spec:
-                    required:
-                      - lastTransitionTime
-                      - status
-                      - type
```
No longer required?

jkyros replied:
Unforunately yes, because in the MCO originally it wasn't: 

https://github.com/openshift/machine-config-operator/blob/886c5c3fc7a96e922a65d83ecd39d19f4921dd50/install/0000_80_machine-config-operator_01_containerruntimeconfig.crd.yaml#L145

JoelSpeed replied:
Ahh right, so in the copy/paste some generation has made it required, and this reverts that? Would be good to squash that change somehow so it's obvious?

Perhaps if we do the copy/paste, then the next commit fixes all the optional vs required, then generate as the final commit, it will be more obvious that there's actually no diff?

jkyros replied:
Thanks, I tried to squish the commits down to a more clear: 
-  add existing -> modify annotations -> generate

hopefully that's...at least a little better?

JoelSpeed replied:
That's what I was hoping for, let me re-review and i'll let you know if anything else can be done :)

JoelSpeed replied:
Ok so these required to optional transitions still exist even when the update comes at the end? Why are these fields switching from required to optional?

JoelSpeed replied:
Can you link to where this CRD currently gets installed into payload to show that these fields aren't correct there? If that's the case, at least we have a record of it then

JoelSpeed replied:
Ahh I see, you've done that already nvm

JoelSpeed commented on machineconfiguration/v1/doc.go:
```diff
@@ -1,5 +1,6 @@
 // +k8s:deepcopy-gen=package,register
 // +groupName=machineconfiguration.openshift.io
 
+// +kubebuilder:validation:Optional
```
Why did you chose to add this? Moving fields from being required to optional could be a breaking change so we need to be careful here

jkyros replied:
Historically..."mistakes were made" https://github.com/openshift/api/pull/1453#discussion_r1186430524. Half of our stuff was generated one way, half of it was generated the other. If you'd rather I "invert" it and go through and tag everything optional that was optional instead, I can do that.

jkyros replied:
Thanks, yep, I have now marked all the fields as optional or required. 

I did keep the `+kubebuilder:validation:Optional` for now because:

-  in the existing MCO we a [field that is marked required](https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/install/0000_80_machine-config-operator_01_containerruntimeconfig.crd.yaml#L51) in our CRD but is `omitempty` [in types.go](https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/pkg/apis/machineconfiguration.openshift.io/v1/types.go#L523) (maintaining CRDs by hand...) and that seemed to be the only way to keep parity. 
-  I did generate both ways (with global optional in doc.go and without) and that field was the only difference


(I don't think preserving it matters *that* much, but I figure we can revisit it next time we make an MCO API change and/or do some "opinionated cleanup" on our API?)

JoelSpeed replied:
Any field that isn't explicitly marked as optional or required, but happens to have `omitempty`, becomes optional by the logic of the generators.

Since the CRD had it as required, it should remain as required

JoelSpeed commented on machineconfiguration/v1/types.go:
```diff
@@ -30,6 +30,10 @@ type ControllerConfig struct {
+	// generated with all fields required, while everything else was generated with optional
+
+	// +kubebuilder:validation:Required
 	// +required
```
This tag isn't correct, please drop the `+required` from everywhere, only use `+kubebuilder:validation:Required`

jkyros replied:
thanks, removed all `+required`

JoelSpeed commented on machineconfiguration/v1/stable.controllerconfig.testsuite.yaml:
```diff
@@ -1,15 +1,101 @@
+          namespace: openshift-config
+        releaseImage: ""
+        rootCAData: Y2VydGlmaWNhdGUK
+        
```
More red markers at the end of files

jkyros replied:
fixed, thanks!

jkyros commented on machineconfiguration/v1/0000_80_containerruntimeconfig.crd.yaml:
```diff
@@ -0,0 +1,141 @@
+  labels:
+    "openshift.io/operator-managed": ""
+  annotations:
+    api-approved.openshift.io: https://github.com/openshift/api/pull/1060
```
nope, that was the previous attempt, I missed one, thanks!

jkyros commented on operator/v1/0000_80_machine-config-operator_01_config.crd.yaml:
```diff
@@ -0,0 +1,192 @@
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    api-approved.openshift.io: https://github.com/openshift/api/pull/TODO(jkyros)
```
fixed, thanks!

jkyros commented on machineconfiguration/v1/stable.controllerconfig.testsuite.yaml:
```diff
@@ -0,0 +1,15 @@
+      apiVersion: machineconfiguration.openshift.io/v1
+      kind: ControllerConfig
+      spec: {}
+      
```
fixed, thanks!

jkyros commented:
Changes are: 
  - rebased
  - squashed out some of the intermediate commits in an attempt to make things clearer
  - removed all usage of `+required` annotation in favor of `+kubebuilder:validation:Required`
  - grouped the machineconfiguration operator stuff together in the same commit 
  - fixed approval annotations to match this PR 
  - fixed test newlines 
  - fixed the capitalization of omitEmpty in StaticPodOperatorStatus in operator types (but we can take it out if we don't want to fix it as part of this)

cdoern commented on machineconfiguration/v1/types.go:
```diff
@@ -167,6 +185,32 @@ type ControllerConfigStatus struct {
+	// +kubebuilder:validation:Required
+	Signer string `json:"signer"`
+
+	// notBefore is the lower boundary for validity
+	// +kubebuilder:validation:Required
+	NotBefore string `json:"notBefore"`
+
+	// notAfter is the upper boundary for validity
+	// +kubebuilder:validation:Required
+	NotAfter string `json:"notAfter"`
```
Not sure that this matters, I added these and for sanity of output string just worked better. I can change if we need to

JoelSpeed replied:
Is this status or spec?

Typically timestamps in kube are all `*metav1.Time` since it has special marshal/unmarshal rules and a standard representation of the time, not sure it's hugely important but had I been doing an API review on this, I would have suggested to be consistent with the other APIs

JoelSpeed replied:
Yeah I think that would be preferable if these fields were introduced this release

jkyros replied:
I've changed these time fields to *metav1.Time, I left it as a separate commit but I can squash it out if need be

cgwalters replied:
It sounds like changing this is tricky for backwards compatibility?  Since controllerconfig is an implementation detail today, I'm not sure we should block on it.

jkyros commented on operator/v1/types_machineconfiguration.go:
whoops, yep, moved it into the operator type commit, thanks

jkyros commented:
that was weird. gateway timeouts? 
/retest

JoelSpeed commented on machineconfiguration/v1/0000_80_controllerconfig.crd.yaml:
```diff
@@ -49,7 +49,6 @@ spec:
               required:
                 - additionalTrustBundle
                 - baseOSContainerImage
-                - baseOSExtensionsContainerImage
```
Removing options from the enum, why has that change come into this PR? Was the CRD out of date in MCO?

jkyros replied:
These specifically in ControllerConfig I did adjust. 

So last time ControllerConfig got regenerated it was a one-off by a "3rd party" PR and those fields got set wrong...and we were apparently in a hurry.  It effectively didn't matter to the MCO because it always supplies the whole structure but we marked it as "probably not great, we should fix it when we migrate the API":  https://github.com/openshift/machine-config-operator/pull/3567/files#r1114291604

And osImageURL is optional now that we're moved to native container format images and that just never got adjusted (again, I think we're supplying the whole thing every time so I don't think it actually matters, but it was wrong).

JoelSpeed replied:
Right, so in this case, you made a breaking change in the 4.13 branch, this is undoing that breaking change, I'm ok with that

JoelSpeed reviewed (commented):
I have concern about the number of fields we are switching from required to optional, what's the justification for switching these fields over? Is it even intentional?

jkyros commented on machineconfiguration/v1/0000_80_kubeletconfig.crd.yaml:
```diff
@@ -134,10 +134,6 @@ spec:
-                    required:
-                      - lastTransitionTime
-                      - status
-                      - type
```
this is currently not required in the MCO: https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/install/0000_80_machine-config-operator_01_kubeletconfig.crd.yaml#L216C1-L216C1

jkyros commented on machineconfiguration/v1/0000_80_machineconfigpool.crd.yaml:
```diff
@@ -87,9 +87,6 @@ spec:
               description: MachineConfigPoolSpec is the spec for MachineConfigPool resource.
               type: object
-              required:
-                - configuration
```
this is currently not required in the MCO: https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/install/0000_80_machine-config-operator_01_machineconfigpool.crd.yaml#L93C13-L93C13

jkyros commented on machineconfiguration/v1/0000_80_machineconfigpool.crd.yaml:
```diff
@@ -220,26 +217,33 @@ spec:
-                - machineCount
-                - readyMachineCount
-                - unavailableMachineCount
-                - updatedMachineCount
```
this is currently not required in the MCO: https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/install/0000_80_machine-config-operator_01_machineconfigpool.crd.yaml#L284

jkyros commented on machineconfiguration/v1/0000_80_machineconfigpool.crd.yaml:
```diff
@@ -220,26 +217,33 @@ spec:
-                      - message
-                      - reason
-                      - status
-                      - type
```
this is currently not required in the MCO https://github.com/openshift/machine-config-operator/blob/40571adccaa3ef49a823abe8d7494bcdc48f2021/install/0000_80_machine-config-operator_01_machineconfigpool.crd.yaml#L293

jkyros reviewed (commented):
For the skew between the base CRD here and the generation after the tags have been adjusted, I believe I can account for all of it

jkyros commented:
Protobufs have been removed :smile:

JoelSpeed commented on machineconfiguration/v1/0000_80_machineconfig.crd.yaml:
```diff
@@ -49,8 +49,6 @@ spec:
           description: "MachineConfig defines the configuration for a machine \n Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer)."
           type: object
-          required:
-            - spec
```
This isn't required in the existing CRD in payload, https://github.com/openshift/machine-config-operator/blob/beca7a6c8a28d83419dd2dfd74e256ca2b80d754/install/0000_80_machine-config-operator_01_machineconfig.crd.yaml#L50-L52C9

JoelSpeed commented:
Apart from https://github.com/openshift/api/pull/1453#discussion_r1288134482, I think everything else I've asked for changes on has been resolved

jkyros commented:
The "certificate time fields" from https://github.com/openshift/api/pull/1453#discussion_r1288134482 have been changed as requested, so...maybe we've got it? :smile: :smile: 

those test failures look systemic and unreleated to what  I did. They passed locally. Let's try again ?
/retest

jkyros commented:
/retest

jkyros commented:
@JoelSpeed we've got green on: 
- https://github.com/openshift/client-go/pull/243
And mostly green on: 
- https://github.com/openshift/machine-config-operator/pull/3747 (My attempt) 
(We stepped on ourselves with the upgrade test because we made that breaking metav1.Time change to our API, but everything else checks out) 
- https://github.com/openshift/machine-config-operator/pull/3840 (Charlie's attempt) 

are we too late to get this in?

JoelSpeed commented:
/lgtm
/hold

When you are ready to merge, go ahead.

If you think this can be merged, with the remaining dependents before the branch gets cut for 4.15, then go ahead, though I'd advise a cautious approach here given how short we are on time

cgwalters commented:
I think ControllerConfig is an implementation detail of the MCO that should absolutely *not* move into openshift/api - we really should aim to remove it entirely because it duplicates lots of data (including secrets) that we really shouldn't be duplicating.

jkyros commented:
> I think ControllerConfig is an implementation detail of the MCO that should absolutely not move into openshift/api - we really should aim to remove it entirely because it duplicates lots of data (including secrets) that we really shouldn't be duplicating.

I did mark it as "internal" at least :smile: 

Until we "get rid of it completely" though, we have to do...something with it. Like we can leave it in the MCO, but it's getting updated there poorly by hand. 

Those embedded objects get updated here, and then don't get vendored back into the MCO all the time, in some cases because folks don't know they need to do that (because the MCO's API is just off by itself) so we're perpetually out of sync.

cgwalters commented:
OK if it's internal only and changing it is hard [let's not change it](https://github.com/openshift/api/pull/1453#discussion_r1295009744) - we have tons of other higher priority tech debt.

JoelSpeed commented:
The new verify-crd-schema check is expected to fail on a PR that is bringing existing types into the API. The idea is that it checks for new violations of CRD rules, but, we accept in certain cases it will need to be overridden. Will override when we are ready to merge

jkyros commented:
Ok cool, thanks @JoelSpeed. 

Are you and/or Eads okay with this merging once the 4.15 branch opens, or do we need to have a longer discussion about ControllerConfig being marked "internal" before we do that?

JoelSpeed commented:
@jkyros Can you add a topic to next weeks control plane arch so that we can discuss there what's required to get this over the line? David had some concerns about the controller config CRD so I think we should make sure those are enumerated

wking commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,747 @@
+const MachineConfigRoleLabelKey = "machineconfiguration.openshift.io/role"
+
+// KubeletConfigRoleLabelPrefix is the label that must be present in the KubeletConfig CR
+const KubeletConfigRoleLabelPrefix = "pools.operator.machineconfiguration.openshift.io/"
```
nit: Perhaps at some point in the future, other custom resources besides KubeletConfigs would use this annotation to associate themselves with pools? Do we want a more generic name for the variable, like `PoolLabelPrefix`?

JoelSpeed replied:
Does this need to be fixed pre-merge?

jkyros replied:
I vote no, it's already been in for a year: https://github.com/openshift/machine-config-operator/pull/3015 and was put in specifically for consumption by the kublet config controller

jkyros replied:
okay, I say that, but I went and looked and...nothing references it right now. I wonder if it ever did?

I still think we should handle this later...but...yeah. Cleanup. :smile:

wking commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,747 @@
+// ControllerConfig describes configuration for MachineConfigController.
+// This is currently only used to drive the MachineConfig objects generated by the TemplateController.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
```
worth putting a "can change at any time for any reason" api into v1alpha1, instead of putting it in `machineconfiguration/v1/`?

jkyros replied:
It probably would if we were just now introducing ControllerConfig. We talked about this in control plane lifecycle arch on 9/19/2023, and my use of level 4 was probably erroneous.  We're just going to mark it level 1 and keep it in v1 since it's already out there.

wking commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,747 @@
+	Network *NetworkInfo `json:"network"`
+}
+
+// ImageRegustryBundle contains information for writing image registry certificates
```
"ImageRegustryBundle" -> "ImageRegistryBundle" (u -> i)

jkyros replied:
good catch, thanks :smile:

wking commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,747 @@
+)
+
+// Network contains network related configuration
+type NetworkInfo struct {
```
Godocs talking about `Network` but the type name is `NetworkInfo`.  Not clear to me what `Info` adds to the name; all of these structures are designed to carry information.  But a common pattern for `*Spec` child types is to be `*Spec` themselves, so `ControllerConfigSpec` would have `NetworkSpec` or, if you wanted to namespace defensively, `ControllerConfigNetworkSpec`.

JoelSpeed replied:
Does this need to be fixed pre-merge?

jkyros replied:
I don't disagree with Trevor philosophically, but I vote no for now, since the fields have been in the MCO for years, and  if they're safe to change now, they're safe to change later once we've thought about it more.

Maybe we can do a pass after this merges of "things we can fix in the MCO API that won't break stuff" and clean this up some? (Yeah, I know we could version the API but ugh, that is unpleasant and I don't want to do it just for icky names  :smile: )

jkyros commented:
Alright, I've: 
- added the missing `InternalRegistryPullSecret` field that the MCO merged the same day we last generated this :smile: 
- fixed the comment typo Trevor found above 
- rebased and regenerated 
- clients have also been regenerated https://github.com/openshift/client-go/pull/243

I notice I'm failing `verify-crd-schema` with even more errors ( `ListsMustHaveSSATags` ) now.

-  Do I need to default those lists to `//+listType=atomic` and we can come back later for merge efficiency gains? Or do we want to merge this to unblock MCO stuff and then come back and address that all at once?

JoelSpeed commented:
> Do I need to default those lists to //+listType=atomic and we can come back later for "list merge efficiency gains"? Or do we want to merge this to unblock MCO stuff and then come back and address that all at once?

Adding the atomic tag is good, as that will show that the field is treated as atomic. But, you can't change this now. The API is out and changing the behaviour of the merge strategy for the list could cause backwards compatibility issues for clients.

If you can add the tags for all the list type failures, we can ignore the other failures as these are "pre-existing unfixable" issues

JoelSpeed commented on operator/v1/types_machineconfiguration.go:
```diff
@@ -0,0 +1,59 @@
+
+	// spec is the specification of the desired behavior of the Machine Config Operator
+	// +kubebuilder:validation:Required
+	// +required
```
Nit, don't need this tag, the kubebuilder version is canonical

jkyros replied:
removed, thanks

JoelSpeed commented on machineconfiguration/v1/types.go:
```diff
@@ -29,7 +29,10 @@ type ControllerConfig struct {
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	// +required
+	// TODO(jkyros): inconsistent historical generation resulted in the controllerconfig CRD being
+	// generated with all fields required, while everything else was generated with optional
```
Is this todo resolved?

jkyros replied:
yes, everything is explicitly tagged now, comment removed

JoelSpeed commented on machineconfiguration/v1/types.go:
```diff
@@ -0,0 +1,752 @@
+
+// ControllerConfigList is a list of ControllerConfig resources
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
```
Should be level 1 right?

jkyros replied:
good catch, I got the main type but I missed the List

JoelSpeed reviewed (commented):
Have a few nits and a few questions about outstanding questions, but I think we are looking very close to go, is that correct?

jkyros commented on machineconfiguration/v1/0000_80_controllerconfig.crd.yaml:
```diff
@@ -0,0 +1,1246 @@
+  labels:
+    "openshift.io/operator-managed": ""
+  annotations:
+    api-approved.openshift.io: https://github.com/openshift/api/pull/1453
```
So that one wasn't previously in the install manifests, the MCO historically applied it when it came up. 

I've added annotations to match the rest of the CRDs:
```
  include.release.openshift.io/ibm-cloud-managed: "true"
  include.release.openshift.io/self-managed-high-availability: "true"
  include.release.openshift.io/single-node-developer: "true"`
```
I didn't see any harm...well...unless the CVO has some sort of "these are annotated but I didn't apply them" cleanup/pruning function I don't know about.

jkyros commented:
Alright so now I've: 
 - added inclusion annotations to the ControllerConfig CRD
 - fixed the ControllerConfigList to be level 1 
 - added `//+listType=atomic ` tags to the lists that we own that verify was crying about (but upon further inspection, a lot of them are from embedded objects and other types here we don't own) 
 - fixed the required tag on the operator field 
 - removed my TODO about the tagging 
 - regenerated clients

I'm thinking "not right now" on Trevor's naming changes -- not that I disagree with him on the rationale, I appreciate the input, but I'm unsure of the blast radius since those aren't new fields (they've been in there for years). If we're gonna improve the naming, I'd rather do it as part of some type of holistic approach (and if they're safe to change now, they should be just as safe to change later).

JoelSpeed commented:
I can't see any issues in net new fields, however, this type has a number of pre-existing failures causing the verify to fail. The verify compares new to old so with an override now should be ok in the future
/override ci/prow/verify-crd-schema 
/lgtm

jkyros commented:
@JoelSpeed are you cool if I release the hold? It's clean, I can safely pull it back into the MCO: https://github.com/openshift/machine-config-operator/pull/3747

JoelSpeed commented:
/hold cancel
